### DISCUSSION TITLE
Is it able to develop a Chrome extension(manifest v3) using frb?

### DISCUSSION THREAD

**Discussion by chenghang6** [fzyzcjy/flutter_rust_bridge#2376] [Q&A]
I tried to turn my flutter app which using frb into a Chrome extension following this article: https://itnext.io/building-a-chrome-extension-with-flutter-a-step-by-step-guide-ea7260c0c452.
I modified manifest.json and index.html as below, and compile it with:
```
flutter_rust_bridge_codegen build-web --release
flutter build web --web-renderer html --csp
```

manifest.json
```
{
    "name": "extension sample",
    "description": "sample",
    "version": "1.0.0",
    "content_security_policy": {
        "extension_pages": "script-src 'self' 'wasm-unsafe-eval'; object-src 'self'"
    },
    "cross_origin_opener_policy": {
        "value": "same-origin"
    },
    "cross_origin_embedder_policy": {
        "value": "require-corp"
    },
    "action": {
        "default_popup": "index.html",
        "default_icon": "icons/Icon-192.png"
    },
    "manifest_version": 3
}
```

index.html
```
<!DOCTYPE html>
<html style="height: 600px; width: 300px">

<head>
  <meta charset="UTF-8" />
  <title>chromextension</title>
</head>

<body>
  <script src="main.dart.js" type="application/javascript"></script>
</body>

</html>
```

There is no problem when running with:
`flutter run -d chrome --web-header=Cross-Origin-Opener-Policy=same-origin --web-header=Cross-Origin-Embedder-Policy=require-corp`

But after loaded into Chrome as an extension, it ran into an error immediately when clicking extension button:
```
core_patch.dart:279 Uncaught Error: EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' 'wasm-unsafe-eval'".

    at new Function (<anonymous>)
    at _web.dart:15:3
    at aih.a (async_patch.dart:303:19)
    at aih.$2 (async_patch.dart:328:23)
    at ahA.$1 (async_patch.dart:279:19)
    at Ph.ai_ (zone.dart:1661:46)
    at Ph.xA (zone.dart:1660:5)
    at ad5.$0 (future_impl.dart:128:29)
    at Object.rS (future_impl.dart:890:13)
    at af.nj (future_impl.dart:657:7)
```

It failed at the very beginning when initing RustLib.
`await RustLib.init();`

**Comment by fzyzcjy** [comment]
Firstly, I wonder whether we can disable that unsafe-eval protection (e.g. by specifying CORS header etc). I am not a web expert so not sure whether we can do that.

Secondly, maybe debug to find which line of code causes trouble. For example, you can do console.error in rust code here and there, and bisect to know which line of code gets stuck.

**Comment by chenghang6** [comment]
Added an error log in my init function. It's not called, no console output.

```
use anyhow::{Ok, Result};
use log::info;
use web_sys::console;

fn log_message(message: &str) {
    console::error_1(&message.into());
}

#[flutter_rust_bridge::frb(init)]
pub fn init_app() -> Result<()> {
    log_message("hello");

    // Default utilities - feel free to customize
    flutter_rust_bridge::setup_default_user_utils();
    Ok(())
}
```

**Comment by chenghang6** [comment]
Added some logs in rust_lib_xxx.js generated by frb codegen.
But the error didn't show after the whole function was executed.
Where can I add more logs?

**Reply by fzyzcjy** [reply]
Maybe try to add `print` here and there in flutter_rust_bridge's Dart code.

**Comment by chenghang6** [comment]
`    at _web.dart:15:3`
Just found that the call stack already shows it's _web.dart:15 which is in frb_dart(frb_dart\lib\src\wasm_module\_web.dart):
```
/// {@macro flutter_rust_bridge.internal}
Future<void> initializeWasmModule({required String root}) async {
  _ensureCrossOriginIsolated();

  final script = web.HTMLScriptElement()..src = '$root.js';
  web.document.head!.append(script);

  await script.onLoad.first;  // line 15

  jsEval('window.wasm_bindgen = wasm_bindgen'); // Or here?

  await _jsWasmBindgen('${root}_bg.wasm').toDart;
}
```

**Reply by fzyzcjy** [reply]
I see. Then you can

```
print('1');
  await script.onLoad.first;  // line 15
print('2');
  jsEval('window.wasm_bindgen = wasm_bindgen'); // Or here?
print('3');
```

e.g. if "1" and "2" is there but "3" is not printed, then we know it is the jsEval line.

**Reply by chenghang6** [reply]
![image](https://github.com/user-attachments/assets/9f1e4d09-8d62-4c6f-bf3a-ba19b2da1ce8)

**Reply by fzyzcjy** [reply]
Great, then we only need to find out a way to run this line of *static* JS code. Maybe we can search about how to do it in Dart.

**Reply by fzyzcjy** [reply]
You can also try to compile your Dart to JS (instead of the new WASM) and see whether this will work.

**Reply by chenghang6** [reply]
> You can also try to compile your Dart to JS (instead of the new WASM) and see whether this will work.

What do you mean "instead of the new WASM"?

**Reply by chenghang6** [reply]
> Great, then we only need to find out a way to run this line of _static_ JS code. Maybe we can search about how to do it in Dart.

Or is it possible to avoid this line of code?

**Reply by fzyzcjy** [reply]
> What do you mean "instead of the new WASM"?

do you compile dart to js or wasm? If the latter, try former; if the former, then ignore this sentence.

> Or is it possible to avoid this line of code?

You can try to comment it out, but it seems to be assigning that variable to global namespace.

**Reply by chenghang6** [reply]
I got a workaround as below for this error.
```
// inject.js
function setWasmBindgen() {
    if (typeof wasm_bindgen !== 'undefined') {
        console.log("found wasm_bindgen, set it globally");
        console.log(wasm_bindgen);
        window.wasm_bindgen = wasm_bindgen;
    } else {
        console.error('wasm_bindgen is not defined');
    }
}
```

```
<!DOCTYPE html>
<html style="height: 600px; width: 300px">

<head>
  <meta charset="UTF-8" />
  <title>chromextension</title>
  <script src="inject.js"></script> <!-- Ensure this is correct -->
</head>

<body>
  <script src="main.dart.js" type="application/javascript"></script>
</body>

</html>
```

Then in _web.dart replace "jsEval('window.wasm_bindgen = wasm_bindgen');" with:
`js.context.callMethod('setWasmBindgen');`

It passed the line of code of "jsEval". But ran into another error:
![image](https://github.com/user-attachments/assets/f7ea108e-270c-42bc-8c38-86d5617b7254)

**Reply by fzyzcjy** [reply]
Great! 

https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/frb_rust/src/third_party/wasm_bindgen/worker_pool.rs#L230

https://github.com/fzyzcjy/flutter_rust_bridge/blob/master/frb_rust/src/misc/web_utils.rs#L24

Then looks like we are having the same problem: js eval in Rust (wasm) code. Given that your workaround works, maybe try to apply it again here and see whether it works at this place.

**Reply by chenghang6** [reply]
The workaround doesn't work here, as this js code is used to get script path.
Using the workaround, it returns the inject script path: chrome-extension://kcdianfnlfcaopcmlebjmjojplkkdann/inject.js.
Which script path is needed here?

**Reply by chenghang6** [reply]
Tested with "flutter run -d chrome".
The script path is http://localhost:59118/pkg/rust_lib_xxxx.js.
For Chrome extension, the path should be: chrome-extension://kcdianfnlfcaopcmlebjmjojplkkdann/pkg/rust_lib_xxxx.js.
I managed to get this path using the workaround.
Then it ran into a bunch of other errors:
![image](https://github.com/user-attachments/assets/5ee79956-646d-4058-b269-71bada80d8ef)
![image](https://github.com/user-attachments/assets/0db00215-7892-47f5-b438-68d2ea0ab1f3)

I guess it's time to give up...

**Reply by fzyzcjy** [reply]
Hmm maybe search or debug about that... 

Again, is it possible to workaround this "wasm-unsafe-eval" thing? For example, is it possible to make your flutter-rust code be in a separate iframe and will it work

**Reply by chenghang6** [reply]
There is nothing related to "wasm-unsafe-eval" currently.

```
    fn spawn(&self) -> Result<Worker, JsValue> {
        let src = &self.script_src;
        let script = format!(
            "importScripts('{}'); // errors here?
            const FRB_ACTION_PANIC = 3;
            onmessage = event => {{
                let init = wasm_bindgen(...event.data).catch(err => {{
                    setTimeout(() => {{ throw err }})
                    throw err
                }})
                onmessage = async event => {{
                    await init
                    const [payload, ...transfer] = event.data
                    try {{
                        wasm_bindgen.receive_transfer_closure(payload, transfer)
                    }} catch (err) {{
                        if (transfer[0] && typeof transfer[0].postMessage === 'function') {{
                            // panic
                            transfer[0].postMessage([FRB_ACTION_PANIC, err.toString()])
                        }}
                        setTimeout(() => {{ throw err }})
                        postMessage(null)
                        throw err
                    }}
                }}
            }}",
            src
        );
        let blob = Blob::new_with_blob_sequence_and_options(
            &Array::from_iter([JsValue::from(script)]).into(),
            BlobPropertyBag::new().type_("text/javascript"),
        )?;
        let url = Url::create_object_url_with_blob(&blob)?;
        let worker: Worker = Worker::new(&url)?;

        // With a worker spun up send it the module/memory so it can start
        // instantiating the wasm module. Later it might receive further
        // messages about code to run on the wasm module.
        let module = wasm_bindgen::module();
        let memory = wasm_bindgen::memory();
        worker.post_message(&Array::from_iter([module, memory]))?;

        Ok(worker)
    }
```

**Reply by fzyzcjy** [reply]
Hmm, but I see the first red message saying refuse to load some scripts. Is that related?

**Reply by chenghang6** [reply]
FRB tries to importScripts("chrome-extension://kcdianfnlfcaopcmlebjmjojplkkdann/pkg/rust_lib_xxxx.js") into web workers.
"importScripts" violates Chrome extension CSP policy.

There are many dynamic imports/evals in FRB. An architecture view is needed to enable support for Chrome extension v3 development.
All we've done is guessing here and guessing there, maybe by luck we can make it work, but can't use it in production.

**Reply by fzyzcjy** [reply]
> All we've done is guessing here and guessing there, maybe by luck we can make it work, but can't use it in production.

No, IMHO we are not guessing, but engineering to work around (or fix) each issue. I say "guess" not because I am guessing (instead of doing science), but because I do not want to say something with too strong assertions, which does not look polite.  If everything works, then it just works and there is usually no worry about in-production issues. Indeed I think it may not be hard for an expert on frontend to fix all these things - but I guess both of us are more Flutter devs than Web devs ;) - so we may need a bit searching and testing.

**Reply by fzyzcjy** [reply]
But yes, I agree if dynamic imports/evals is forbidden, then frb needs some refactors before it can be used. That's why I say above, is there a way to allow such dynamic import/eval?

**Reply by chenghang6** [reply]
> But yes, I agree if dynamic imports/evals is forbidden, then frb needs some refactors before it can be used. That's why I say above, is there a way to allow such dynamic import/eval?

Yes, that's what I meant, sorry for seemed unpolite.
We need to know clearly what Chrome extension manifest v3 forbids and make some rules if planning to support it, or even we make it work now, new code in the future may break it.

**Reply by fzyzcjy** [reply]
It's OK, no worries!

> We need to know clearly what Chrome extension manifest v3 forbids 

Good point. Indeed I am wondering whether it is because of the specific webpage you are targeting that causes this, instead of the limitation of extension.
